drivers/video/gc2145.c

```diff
@@ -0,0 +1,1112 @@
+				uint8_t value)
+{
+	uint8_t tries = 3;
+
+	/**
+	 * It rarely happens that the camera does not respond with ACK signal.
+	 * In that case it usually responds on 2nd try but there is a 3rd one
+	 * just to be sure that the connection error is not caused by driver
+	 * itself.
+	 */
+	while (tries--) {
+		if (!i2c_reg_write_byte_dt(spec, reg_addr, value)) {
+			return 0;
+		}
+		/* If writing failed wait 5ms before next attempt */
+		k_msleep(5);
+	}
+	LOG_ERR("failed to write 0x%x to 0x%x", value, reg_addr);
+
+	return -1;
+}
+
+static int gc2145_read_reg(const struct i2c_dt_spec *spec, uint8_t reg_addr)
+{
+	uint8_t tries = 3;
+	uint8_t value;
+
+	/**
+	 * It rarely happens that the camera does not respond with ACK signal.
+	 * In that case it usually responds on 2nd try but there is a 3rd one
+	 * just to be sure that the connection error is not caused by driver
+	 * itself.
+	 */
+	while (tries--) {
+		if (!i2c_reg_read_byte_dt(spec, reg_addr, &value)) {
+			return value;
+		}
+		/* If reading failed wait 5ms before next attempt */
+		k_msleep(5);
+	}
+	LOG_ERR("failed to read 0x%x register", reg_addr);
+
+	return -1;
+}
+
+static int gc2145_write_all(const struct device *dev,
+				const struct gc2145_reg *regs, uint16_t reg_num)
+{
+	uint16_t i = 0;
+	const struct gc2145_config *cfg = dev->config;
+
+	for (i = 0; i < reg_num; i++) {
+		int err;
+
+		err = gc2145_write_reg(&cfg->i2c, regs[i].addr, regs[i].value);
+		if (err) {
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static int gc2145_soft_reset(const struct device *dev)
+{
+	int ret = 0;
+	const struct gc2145_config *cfg = dev->config;
+
+	/* Initiate system reset */
+	ret |= gc2145_write_reg(&cfg->i2c, GC2145_REG_RESET, GC2145_REG_SW_RESET);
+
+	k_msleep(30);
+
+	return ret;
+}
+
+static int gc2145_set_ctrl_vflip(const struct device *dev, int enable)
+{
+	const struct gc2145_config *cfg = dev->config;
+
+	/* Using the Analog mode 1 register (0x17) */
+	int ret = gc2145_read_reg(&cfg->i2c, GC2145_REG_AMODE1);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Set the vertical flip state */
+	uint8_t old_value = ret;
+
+	ret = gc2145_write_reg(&cfg->i2c, GC2145_REG_AMODE1,
+						(old_value & 0xFC) |
+						(enable << 1));
+	return ret;
+}
+
+static int gc2145_set_ctrl_hmirror(const struct device *dev, int enable)
+{
+	const struct gc2145_config *cfg = dev->config;
+
+	/* Using the Analog mode 1 register (0x17) */
+	int ret = gc2145_read_reg(&cfg->i2c, GC2145_REG_AMODE1);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Set the vertical flip state */
+	uint8_t old_value = ret;
+
+	ret = gc2145_write_reg(&cfg->i2c, GC2145_REG_AMODE1,
+						(old_value & 0xFC) |
+						enable);
+	return ret;
+}
+
+static int gc2145_set_window(const struct device *dev,
+							uint16_t reg,
+							uint16_t x,
+							uint16_t y,
+							uint16_t w,
+							uint16_t h)
+{
+	const struct gc2145_config *cfg = dev->config;
+	int ret = 0;
+
+	/* P0 regs */
+	ret |= gc2145_write_reg(&cfg->i2c, 0xFE, 0x00);
```

Comment:
don't do these partial failure handling things please, if the register write fail the i2c bus is probably hosed, no point trying to push forward with a partial failure, just check every call and return the error at the first that fails